### ISSUE TITLE
spyExisting(Object o): Turn an existing object into a spy

### ISSUE THREAD

**Issue by moltmann** [mockito/mockito#1277]
**tl:dr** I want to add a new API to make an existing object a spy. One use case is spying on Android Activity-objects: These objects provide access to Android's system services. It would very useful for testing to replace a system service with a special testing service.

_This is about mocking on Android devices. If you can fully mock the environment you should run your test on your workstation. Mocking on Android is usually partial mocking._

In Android most interaction to the system have the following pattern:

    public void onClick() {
        PrintManager pm = ((Activity)this).getSystemService(PrintManager.class);
        List printJobs = pm.getPrintJobs();
        ...
    }

Fortunately Androids system services are _mostly_ independent so that mocking one service and leaving the other services behaving normally is a viable option. It would be nice if you could mock out the `T Activity.getSystemService(Class<T> c)` method and return your mock system service.

You could inject mocked behavior by modifying the production code and routing all `getSystemService` calls though a mocking-system. This would make the production code harder to read only to deal with a limitation of mockito.

Currently `spy()` always spies on a wrapper of the object. It would be nice if you could spy on an existing Activity object, e.g.

In the test method:
1. Register for ActivityLifecycleCallbacks
2. Start activity
3. Wait until you get the callback that the activity is created
4. Make the existing activity-object a spy
5. Mock getSystemService for your activity to replace `getSystemService` to return your test-service
6. Trigger operations on the activity

The API would be very simple

    public class Mockito {
        static void spyExisting(Object objectToSpyOn) { ... }
        ...
    }

**Comment by Brianhedgecock** [comment]
all"o" spy.

**Comment by mockitoguy** [comment]
Hey @moltmann, thank you for reporting!

>Currently spy() always spies on a wrapper of the object.

Can you elaborate why the existing method ```Mockito.spy(object)``` does not work for you? This would really help us understand the use case!

**Comment by moltmann** [comment]
Simplifying a little bit:
- Usually in Android there is no continuously running method, everything is callback based.
- In Android each screen is usually internally represented by a different subclass of the Activity class. The behavior of this screen is defined by overriding the methods of the Activity class.
- One set of important methods of Activity are the lifecycle methods. E.g. `Activity#onCreate` is called when the screen is shown. In this method the app created the UI and registers callbacks for button presses.
- As the activity is created by the system `Activity#onCreate` is often the first code of an app that it run. The constructor of your Activity class is called by the system, never manually.
- The app accesses the rest of the system by calling `Activity#getSystemService(Class yourFavoriteSystemService)`.

Hence to fake the behavior of the system you usually have to stub the `Activity#getSystemService` method on an object that the test does not create itself. As the test does not call the constructor of the activity there is no opportunity to spy via cloning. Also only the original activity object is registered with the system, hence only it will get callbacks.

Here is an example:

```
@RunWith(AndroidJUnit4.class)
public class PermissionGrantLocationLoggerTest {
    // Ask the system to start the MySubclassOfActivity activity **before** the test is run
    @Rule
    public ActivityTestRule<MySubclassOfActivity> activityRule = new ActivityTestRule<>(MySubclassOfActivity.class);

    @Test
    public void test() throws Exception {
MySubclassOfActivity myActivity = activityRule.getActivity();
        spyOn(myActivity);

        // How you could intercept communication of the app code with the platform.
        // E.g. this intercepts communication of the app with the Location-subssystem
       when(myActivity.getSystemService(LocationManager.class)).thenReturn(mock(LocationManager.class));

       // Now use a UI-automation framework such as Espresso to click a button that triggers code that accesses the location subsystem
       ...
     }
}
```

A complete example can be found here:
- [App code](https://github.com/moltmann/dexmaker-mockito-inline-extended-examples/blob/master/app/src/main/java/priv/moltmann/locationlogger/LocationLogger.java)
- [Test code](https://github.com/moltmann/dexmaker-mockito-inline-extended-examples/blob/master/app/src/androidTest/java/priv/moltmann/locationlogger/PermissionGrantLocationLoggerTest.java)

**Comment by wpc009** [comment]
No nothing after two and half years.

**Comment by wpc009** [comment]
> object



> Hey @moltmann, thank you for reporting!
> 
> > Currently spy() always spies on a wrapper of the object.
> 
> Can you elaborate why the existing method `Mockito.spy(object)` does not work for you? This would really help us understand the use case!

The current `Mockito.spy(object)` create a new mocked instance out of the original instance. It does not affect the reference to the original instance. So, `mocking` and `verify` will not working unless you replace all the reference of the original instance.
